Specify mksquashfs path during configure

jmstover opened the pull request:
**Description of the Pull Request (PR):**

Add a `--with-mksquashfs` configure option to allow the user to specify the full path where `mksquashfs` is located.

For example:
`./configure --with-mksquashfs=/usr/local/mksquashfs/mksquashfs`

**This fixes or addresses the following GitHub issues:**

- Ref: #1523 

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

cclerget committed:
fileput and container_mkpath wrapper for both privilege and non privileged operations

cclerget committed:
Call fd_cleanup after container setup is done.
Add a callback function for fd_cleanup to customize
cleanup both for action and start code.

cclerget committed:
Check if user identify can't be obtain and avoid segmentation fault

DrDaveD committed:
allow --home to still work with "mount home = no"

GodloveD committed:
Merge pull request #1490 from cclerget/development-2.x-issue-1489

Call fd_cleanup after container setup

GodloveD committed:
Merge pull request #1491 from cclerget/development-2.x-fix-2.5

Fix permission denied on NFS root_squash when changing directory

cclerget committed:
Fix debootstrap issue

cclerget committed:
Add deboostrap and yum definition file tests

GodloveD committed:
Merge pull request #1500 from cclerget/release-2.5-debootstrap-fix

Fix debootstrap issue

jmstover committed:
Fix small typo

GodloveD committed:
updating files for 2.5.1

jmstover committed:
Merge pull request #1506 from jmstover/NNP_Fix

Define PR_(S|G)ET_NO_NEW_PRIVS ourself

GodloveD committed:
Merge pull request #1509 from GodloveD/cherry-pick-1506

Merge pull request #1506 from jmstover/NNP_Fix

bauerm97 committed:
Merge pull request #1502 from jmstover/filebind_typo

Fix small typo

cclerget committed:
Remove open check from singularity_mount

jmstover committed:
Check GID in 45-targetmode.sh

jmstover committed:
Add --with-mksquashfs option to specify mksquashfs path

jmstover committed:
Clean up output a bit

jmstover committed:
Add /bin to mksquashfs path output

jmstover committed:
Add fall back check for mksquashfs

jmstover committed:
Change build.exec to add MKSQUASHFS_PATH to PATH if set

GodloveD committed:
Merge pull request #1525 from cclerget/development-2.x-mount-fix

Fix mount invalid argument and restrictive directory binds

GodloveD committed:
Revert "Fix mount invalid argument and restrictive directory binds"

GodloveD committed:
Merge pull request #1531 from singularityware/revert-1525-development-2.x-mount-fix

Revert "Fix mount invalid argument and restrictive directory binds"

cclerget committed:
Remove open check from singularity_mount

GodloveD committed:
Merge pull request #1532 from cclerget/release-2.5-mount-fix

Remove open check from singularity_mount

cclerget committed:
Display errno string for realpath for debug help

GodloveD committed:
Merge pull request #1535 from cclerget/release-2.5-strerror

Display errno string for realpath for debug purpose

DrDaveD committed:
build centos:7 for now without yum-builddep

boegel commented on configure.ac:
```diff
@@ -266,9 +266,32 @@ AC_CHECK_FUNCS(setns, [
+           )
+AS_IF([test "x$with_mksquashfs" != "x"],
+      [
+       AC_CHECK_PROG([MKSQUASHFS_CHECK],[mksquashfs],[yes],[no],[$with_mksquashfs/bin])
```
By default `mksquashfs` isn't installed into a `bin` subdirectory, so maybe also check in `$with_mksquashfs/` directly? See https://github.com/easybuilders/easybuild-easyconfigs/pull/6287/files/4d84fcea3940a4c6afbd45722a590f83168801c7#diff-cefe1a16fe4b9a2e7bcdec604d37482e

jmstover replied:
I'll need to build it ... but don't you do something like:

`./configure --prefix=/some/path`

Where `/some/path` is the root? Then `--bindir` is `/some/path/bin`? 

So from your test example above, prefix for the `mksquashfs` build would be:
`./configure --prefix=${EBROOTSQUASHFS}`

boegel replied:
There's no `configure` script, only a `Makefile`, you can control where it gets installed via `make install INSTALL_DIR=...`, cfr. https://github.com/plougher/squashfs-tools/blob/master/squashfs-tools/Makefile#L123

jmstover replied:
Ahh... it's one of those. Fun!

I have to keep the assumption there's going to be a `bin/`.  Generally, the `with` options that expect a directory are the _root_ (`--prefix`) where it was installed, and assuming a autoconf configuration, there will be a `$prefix/bin`.

If someone else doesn't like it they can put in a PR to change it, or @GodloveD can change it... but, for me... I can't make myself do it.

Dave: If you think it would be good to change... the best way would probably be to force the entire path to the `mksquashfs` binary when the `--with-mksquashfs` option is used.

boegel replied:
That last idea actually makes more sense to me...

jmstover replied:
> That last idea actually makes more sense to me...

Yeah... but it breaks how the `--with-*` options are expected (and it totally doesn't make more sense to me after years of putting the "root" location in when using it =) ).

We aren't doing anything fancy so it would probably work fine, but IIRC, some of the macros expect the given directory there to be the *root* location (i.e. the `--prefix` value).

GodloveD replied:
Yeah, I agree with @boegel here that we can't assume a `/bin`.  Especially if `mksquashfs` is not installed in a `/bin` directory by default.  Forcing the full path sounds like a good idea to me.  Want to give it a shot and see what it looks like?

boegel commented on configure.ac:
```diff
@@ -266,9 +266,32 @@ AC_CHECK_FUNCS(setns, [
+              AC_MSG_RESULT([using $with_mksquashfs/bin])
+             ],
+             [
+              AC_MSG_WARN([mksquashfs not found - needed at runtime for full build functionality])
```
Shouldn't this be fatal?

jmstover replied:
No. mksquashfs is only needed when you build an image.  So, we're just throwing out a warning that it's needed for full "build functionality". The host you're _building_ on may not have it... but that may also not be the location where you're going to be running something like:  `singularity build [blah]`  -- from.

DrDaveD committed:
remove yum-builddep step, it does not do anything

GodloveD commented on libexec/cli/build.exec.in:
```diff
@@ -36,6 +36,11 @@ else
 
+OLDPATH=${PATH}
+if [ -n "@MKSQUASHFS_PATH@" ]; then
+    PATH=${PATH}:@MKSQUASHFS_PATH@
```
I would prefer if we just forced the full path instead of changing the PATH and then replacing.  Something more like what I did here.  https://github.com/singularityware/singularity/pull/1094/files

jmstover committed:
Use full path to binary for --with-mksquashfs option

Add ignore for libexec/cli/build.exec

GodloveD commented on libexec/cli/build.exec.in:
```diff
@@ -36,12 +36,9 @@ else
 
-if ! singularity_which mksquashfs > /dev/null 2>&1; then
+if ! singularity_which ${MKSQUASHFS:-mksquashfs} > /dev/null 2>&1; then
     message ERROR "You must install squashfs-tools to build images\n"
```
Can we expand this error to tell the user that they must either install squashfs-tools or point to the correct location at build time?  (Phrased better of course.)

jmstover committed:
dd status=none unknown

Hack to work around older dd binaries where `status=LEVEL` is not
known. To my knowledge, these are all `status=noxfer`

cclerget committed:
Fix passwd abort when user identity not found

cclerget committed:
Merge pull request #1580 from cclerget/release-2.5-passwd-fix

Fix passwd abort when user identity not found

jmstover committed:
Merge pull request #1556 from jmstover/pull_debug_fix

Pull --debug fix on docker:// URI

jmstover committed:
Merge pull request #1572 from jmstover/rel25_pr1560

Fix status=none dd error

jmstover committed:
Merge pull request #1528 from jmstover/targetmode_gid_fix

Check GID in 45-targetmode.sh

jmstover committed:
Expand on error text if mksquashfs isn't found

GodloveD commented on libexec/cli/build.exec.in:
```diff
@@ -36,8 +36,13 @@ else
-    message ERROR "You must install squashfs-tools to build images\n"
+MKSQUASHFS="@MKSQUASHFS_PATH@"
+
+if ! singularity_which ${MKSQUASHFS:-mksquashfs} > /dev/null 2>&1; then
```
Sorry, can you also edit the warning in Makefile.in.

GodloveD reviewed (changes requested):
It works for me if I test it, but I'm not seeing the correct warning when I build from source.  Can you add the same lines into the Makefile.in too?

jmstover committed:
Merge branch 'mksquashfs_path' of github.com:jmstover/singularity into mksquashfs_path

jmstover committed:
Update toplevel Makefile.in
